examples/todoapp/lib/base/repositories/todo_repository.dart

```diff
@@ -1,45 +1,212 @@
-        () => dataSource.updateCompletedForAll({'completed': completed}),
+        () async {
+          final result =
+              await dataSource.updateCompletedForAll({'completed': completed});
+          await localDataSource.updateCompletedForAll(completed);
+          return result;
+        },
+      ).onError(
+        (error, stackTrace) => handleError(
+          error,
+          localDataSource.updateCompletedForAll(
+            completed,
+            synced: false,
+            action: TodoModelActions.update.name,
+          ),
+        ),
       );
 
-  Future<List<TodoModel>> deleteCompleted() => _errorMapper.execute(
-        () => dataSource.deleteCompleted(),
+  Future<List<$TodoModel>> deleteCompleted() => _errorMapper.execute(
+        () async {
+          final response = await dataSource.deleteCompleted();
+          await localDataSource.deleteCompleted();
+          return response;
+        },
+      ).onError(
+        (error, stackTrace) => handleError(
+          error,
+          localDataSource.softDeleteCompleted(
+            synced: false,
+            action: TodoModelActions.delete.name,
+          ),
+        ),
       );
 
   Future<void> deleteTodoById(String id) => _errorMapper.execute(
-        () => dataSource.deleteTodoById(id),
-      );
+        () async {
+          await dataSource.deleteTodoById(id);
+          await localDataSource.deleteTodoById(id);
+        },
+      ).onError((error, stackTrace) {
+        if (error is NoConnectionErrorModel) {
+          localDataSource.softDeleteTodoById(
+            id,
+            synced: false,
+            action: TodoModelActions.delete.name,
+          );
+          return;
+        }
+        if (error is ErrorModel) {
+          throw error;
+        } else {
+          throw UnknownErrorModel();
+        }
+      });
 
-  Future<TodoModel> fetchTodoById(String id) => _errorMapper.execute(
-        () => dataSource.getTodoById(id),
+  Future<$TodoModel> fetchTodoById(String id) => _errorMapper.execute(
+        () async {
+          final result = await dataSource.getTodoById(id);
+          return result;
+        },
+      ).onError((error, stackTrace) => handleError(
+            error,
+            localDataSource.getTodoById(id),
+          ));
+
+  Future<void> syncronize() async {
+    final List<TodoModel> unsyncedTodos =
+        await localDataSource.getAllUnsyncedTodos();
+
+    if (unsyncedTodos.isNotEmpty) {
+      await _errorMapper.execute(
+        () async {
+          await localDataSource.unpauseSync();
+          final result = await dataSource.syncTodos({'todos': unsyncedTodos});
+          await localDataSource.deleteMany(unsyncedTodos);
+          await localDataSource.addMany(result);
+        },
       );
+    }
+  }
+
+  Future<T> handleError<T>(
+    Object? error,
+    Future<T> function,
+  ) {
+    if (error is NoConnectionErrorModel) {
+      return function;
+    }
+    if (error is ErrorModel) {
```

Comment:
The responsibility of the ErrorMapper class is to ... map an exception to a business error model so there is no need of checking the error here and to throw a `UnknownErrorModel` in case the exception is not of type ErrorModel.

With that being said feel free to simplify the method to execute the `function` callback if the error is NoConnectionErrorModel and rethrow the exception otherwise

Reply:
Fixed